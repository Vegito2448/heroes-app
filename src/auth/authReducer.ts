
export interface AuthState {
  logged?: boolean;
  name?: string;
}

export type AuthAction =
  | { type: 'logout'; }
  | { type: 'login', payload?: AuthState; };

export const authReducer = (state: AuthState, action: AuthAction): AuthState => {
  switch (action.type) {
    case 'login':
      return {
        ...action.payload,
        logged: true,
      };
    case 'logout': return {
      logged: false,
      name: '',
    };
    default:
      return state;
  }
};