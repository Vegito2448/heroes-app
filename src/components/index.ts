export * from './heroes';
export * from './ui';
export * from './WelcomePage';
