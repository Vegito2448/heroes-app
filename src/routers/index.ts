export * from './router';
export * from './routesConfig';
