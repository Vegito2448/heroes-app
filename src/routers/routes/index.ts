export * from './DcScreen';
export * from './ErrorPage';
export * from './HeroScreen';
export * from './Login';
export * from './MarvelScreen';
export * from './PrivateRoute';
export * from './PublicRoute';
export * from './Root';
export * from './SearchScreen';
