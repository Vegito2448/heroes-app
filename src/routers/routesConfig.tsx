import queryString from "query-string";
import { LoaderFunctionArgs, redirect } from "react-router-dom";
import { getHeroById, getHeroes, getHeroesByPublisher } from "../api";
import { DcScreen, HeroScreen, MarvelScreen, PrivateRoute, SearchScreen } from "./routes";

const createLoaderForPublisher = (publisher: string) => async () => {
  const heroes = await getHeroesByPublisher({ publisher });
  return { heroes };
};

const routesConfig = [
  {
    path: 'marvel',
    loader: createLoaderForPublisher("Marvel Comics"),
    element: <PrivateRoute children={<MarvelScreen />} />,
    title: "Marvel Heroes"
  },
  {
    path: 'dc',
    loader: createLoaderForPublisher("DC Comics"),
    element: <PrivateRoute children={<DcScreen />} />,
    title: "DC Heroes"
  },
  {
    path: "hero/:heroId",
    element: <PrivateRoute children={<HeroScreen />} />,
    loader: async ({ params: { heroId: id } }: LoaderFunctionArgs) => {
      if (isNaN(Number(id)) || !id) {
        return redirect('/');
      }


      const hero = await getHeroById({
        id
      });

      if (hero instanceof Error || !hero) {
        return redirect('/');
      }

      return {
        hero
      };
    },
  },
  {
    path: 'search',
    element: <PrivateRoute children={<SearchScreen />} />,
    loader: async ({ request }: LoaderFunctionArgs) => {
      const { url } = request;

      const { q = '' } = queryString.parse(url.split('?')[1]);

      if (!q) {
        return {};
      }

      const allHeroes = await getHeroes();

      const heroes = Array.isArray(allHeroes) ? allHeroes.filter(({ name }) => name.toLowerCase().includes(String(q).toLowerCase())) : allHeroes;

      return {
        heroes
      };
    },
    title: "Search Heroes"
  }
];
export { routesConfig };

