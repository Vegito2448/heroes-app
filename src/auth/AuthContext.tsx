import { createContext, Dispatch, ReactNode, useEffect, useReducer } from "react";
import { AuthAction, authReducer, AuthState } from "./authReducer";

interface AuthContextProps {
  user: AuthState;
  dispatch: Dispatch<AuthAction>;
}

const initialState: AuthState = {
  logged: false,
  name: ''
};

const init = () => JSON.parse(localStorage.getItem('user') as string) || initialState;


const AuthContext = createContext<AuthContextProps>({
  user: {
    logged: false
  },
  dispatch: () => { }
});

const AuthProvider = ({ children }: {
  children: ReactNode;
}) => {

  const [user, dispatch] = useReducer(authReducer, initialState, init);

  useEffect(() => {
    localStorage.setItem('user', JSON.stringify(user));
  }, [user]);

  return (
    <AuthContext.Provider value={{
      user,
      dispatch
    }}>
      {children}
    </AuthContext.Provider>
  );
};

export { AuthContext, AuthProvider };

