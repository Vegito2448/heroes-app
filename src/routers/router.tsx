import {
  createBrowserRouter,
  RouteObject
} from "react-router-dom";
import { ErrorPage, Login, PublicRoute, Root } from "./routes";
import { routesConfig } from "./routesConfig";


export const allRoutes: RouteObject[] = [
  {
    path: "/login",
    element: <PublicRoute children={<Login />} />,
  },
  {
    path: "/",
    element: <Root />,
    children: routesConfig as RouteObject[],
    errorElement: <ErrorPage />
  },
];

export const router: ReturnType<typeof createBrowserRouter> = createBrowserRouter(allRoutes);

