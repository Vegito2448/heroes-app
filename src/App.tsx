import { RouterProvider } from "react-router-dom";
import { AuthProvider } from "./auth";
import { router } from './routers';

function App() {
  return (
    <AuthProvider children={<RouterProvider router={router} />} />
  );
}

export default App;
