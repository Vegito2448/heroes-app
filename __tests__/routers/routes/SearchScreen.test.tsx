import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { createMemoryRouter, RouterProvider } from "react-router-dom";
import { AuthContext } from "../../../src/auth";
import { allRoutes } from "../../../src/routers";

describe('Search Screen Component Test', () => {// Test for rendering the component

  const heroToSearch = 'batman';

  const paramWithHero = `?q=${heroToSearch}`;

  let router: ReturnType<typeof createMemoryRouter>;

  beforeEach(() => {
    router = createMemoryRouter(allRoutes, {
      initialEntries: ['/search'],
      initialIndex: 0
    });

    userEvent.setup();

    render(
      <AuthContext.Provider
        value={{
          user: {
            name: 'test',
            logged: true
          },
          dispatch: vi.fn()
        }}
      >
        <RouterProvider router={router} />
      </AuthContext.Provider>
    );
  });

  test('should render without crashing', async () => {

    const search = await screen.findByRole('heading', { name: /search screen/i });

    expect(search).toBeInTheDocument();

  });

  // Test for search form interaction
  test('should update URL parameters on form submit', async () => {

    const searchInput = await screen.findByRole('textbox');

    const searchButton = await screen.findByRole('button', { name: /Search.../i });

    expect(searchButton).toBeDisabled();

    await userEvent.type(searchInput, heroToSearch);

    expect(searchInput).toHaveValue(heroToSearch);

    expect(searchButton).toBeEnabled();

    await userEvent.click(searchButton);

    expect(router.state.navigation.location?.search).toBe(paramWithHero);

    const batmanCard = await screen.findByRole('heading', { name: /terry mcginnis/i });

    expect(batmanCard).toBeInTheDocument();

  });



  // Test for displaying "Search a hero" message
  test('should display "Search a hero" message when no search has been made', async () => {

    screen.debug();

    const searchInput = await screen.findByRole('textbox');

    const searchButton = await screen.findByRole('button', { name: /Search.../i });

    const notValue = 'not a hero';

    await userEvent.type(searchInput, notValue);

    expect(searchInput).toHaveValue(notValue);

    await userEvent.click(searchButton);

    const notHeroMessage = await screen.findByText(/No heroes found/i);

    expect(notHeroMessage).toBeInTheDocument();

  });

});