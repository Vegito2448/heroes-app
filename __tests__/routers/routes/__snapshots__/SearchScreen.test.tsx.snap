// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Search Screen Component Test > should match the snapshot 1`] = `<div />`;
